### PR TITLE
peer sharing changes

### PR DESCRIPTION
**Pull request by coot**
# Description

- **netsim: fixed a typo**
- **ConnectionId: changed Ord instance**
- **ConnectionId: use named field puns**
- **peer-sharing: protocol violation**
- **connection-manager: manage its own StdGen**
- **inbound-governor: (re)added PublicInboundGovernorState**
- **inbound-governor: updated tests**
- **inbound-governor: verify inbound governor state invariant**
- **inbound-governor: added a TODO note**
- **diffusion: refactored order of calls**
- **outbound-governor: renamed fuzzRng to stdGen**
- **outbound-governor: replaced readNewInboundConnection with readInboundPeers**
- **outbound-governor: light peer sharing using readInboundPeers API**
- **outbound-governor: randomness**
- **outbound-governor: fixed a counterexample label**
- **Updated CHANGELOG files**

# Checklist

### Quality
* [x] Commit sequence makes sense and have useful messages, see [ref][contrib#git-history].
* [x] New tests are added and existing tests are updated.
* [x] Self-reviewed the PR.

### Maintenance
* [x] Linked an [issue][link-issue] or added the PR to the current sprint of [`ouroboros-network`][project] project.
* [x] Added labels.
* [x] Updated changelog files.
* [ ] The documentation has been properly updated, see [ref][contrib#documentation].

[project]: https://github.com/orgs/IntersectMBO/projects/5/views/1
[link-issue]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=
[contrib#git-history]: https://github.com/IntersectMBO/ouroboros-network/blob/master/CONTRIBUTING.md#git-history
[contrib#documentation]: https://github.com/IntersectMBO/ouroboros-network/blob/master/CONTRIBUTING.md#documentation

### CODE REVIEW

File: ouroboros-network-framework/src/Ouroboros/Network/InboundGovernor.hs [hunk 1]
```diff
@@ -155,7 +173,17 @@ inboundGovernor trTracer tracer inboundInfoChannel
                  )
                  (igsConnections state)
             <> FirstToFinish (
-                 NewConnection <$> InfoChannel.readMessage inboundInfoChannel)
+                 NewConnection <$> InfoChannel.readMessage inboundInfoChannel
+               )
+            <> FirstToFinish  (
+                 -- move connections from
```
Comment (bolt12) [inline ouroboros-network-framework/src/Ouroboros/Network/InboundGovernor.hs:181]: unfinished comment?

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs [hunk 2]
```diff
@@ -604,8 +605,9 @@ peerSelectionGovernorLoop tracer
     evalGuardedDecisions :: Time
                          -> PeerSelectionState peeraddr peerconn
                          -> m (TimedDecision m peeraddr peerconn)
-    evalGuardedDecisions blockedAt st =
-      case guardedDecisions blockedAt st of
+    evalGuardedDecisions blockedAt st = do
+      inboundPeers <- readInboundPeers actions
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs:609]: should this be moved inside `KnownPeers.belowTarget` ?

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs [hunk 3]
```diff
@@ -1381,6 +1383,30 @@ pickPeers PeerSelectionState{localRootPeers, publicRootPeers, knownPeers}
              ++ " which is outside of the set given to pick from"
 
 
+-- | Pick some known peers.
+--
+pickPeers :: (Ord peeraddr, Functor m, HasCallStack)
+          => PeerSelectionState peeraddr peerconn
+          -> PickPolicy peeraddr m
+          -> Set peeraddr -> Int -> m (Set peeraddr)
+{-# INLINE pickPeers #-}
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs:1398]: Why is this needed?

**Review by bolt12 — APPROVED** [review]
LGTM! Just some very minor comments, I'll approve

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs (hunk 2 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs:609]: I tried it, but then it's a bit tricky to reason that we don't stale peer sharing when there are no inbound peers; the same RNG might be used when an STM transaction is retried.  I concluded we won't but the argument was subtle, and this is much more explicit, since we have `inboundPeers` when a decision is made.

I made the `readInboundPeers` quite efficient; the value inside of it is computed incrementally so has good amortised bounds; and reading is also cheap (`O(n)` since it's using `Map.map`).

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs (hunk 3 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs:1398]: Probably not, it's just a tiny wrapper which likely will be inlined anyway.

File: ouroboros-network-framework/src/Ouroboros/Network/InboundGovernor.hs (hunk 1 above)
Comment (coot) [reply ouroboros-network-framework/src/Ouroboros/Network/InboundGovernor.hs:181]: hmm, maybe it was cut during a rebase, or just unfinished.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs (hunk 2 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs:609]: We could even make `readInboundPeers` `O(1)` (amortised); we just need to pass `versionData -> DataFlow` to `inbound-governor`.

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs:609]: But I thought that maybe in the future, we will need a bit more data from `versionData`, so I left it as is.

Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs:609]: I see, that makes sense! Thank you for explaining!
